Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -404,107 +352,72 @@ void CGPathAddArc(CGMutablePathRef path,
     }
 }
 
 /**
  @Status Interoperable
 */
-void CGPathAddRect(CGMutablePathRef path, const CGAffineTransform* m, CGRect rect) {
+void CGPathAddRect(CGMutablePathRef path, const CGAffineTransform* transform, CGRect rect) {
     if (path == NULL) {
         return;
     }
 
-    CGPathMoveToPoint(path, m, CGRectGetMinX(rect), CGRectGetMinY(rect));
+    CGPathMoveToPoint(path, transform, CGRectGetMinX(rect), CGRectGetMinY(rect));
 
-    CGPathAddLineToPoint(path, m, CGRectGetMaxX(rect), CGRectGetMinY(rect));
-    CGPathAddLineToPoint(path, m, CGRectGetMaxX(rect), CGRectGetMaxY(rect));
-    CGPathAddLineToPoint(path, m, CGRectGetMinX(rect), CGRectGetMaxY(rect));
+    CGPathAddLineToPoint(path, transform, CGRectGetMaxX(rect), CGRectGetMinY(rect));
+    CGPathAddLineToPoint(path, transform, CGRectGetMaxX(rect), CGRectGetMaxY(rect));
+    CGPathAddLineToPoint(path, transform, CGRectGetMinX(rect), CGRectGetMaxY(rect));
     CGPathCloseSubpath(path);
 }
 
 /**
- @Status Interoperable
+ @Status Caveat
+ @Notes Ignores Affine Transform
 */
-void CGPathAddPath(CGMutablePathRef path, const CGAffineTransform* m, CGPathRef toAdd) {
+void CGPathAddPath(CGMutablePathRef path, const CGAffineTransform* transform, CGPathRef toAdd) {
     if (path == NULL || toAdd == NULL) {
         return;
     }
 
-    CGPathRef pathObj = path;
-    CGPathRef copyObj = toAdd;
-
-    if (pathObj->_count + copyObj->_count >= pathObj->_max) {
-        pathObj->_max += copyObj->_count;
-        pathObj->_elements = (CGPathElementInternal*)IwRealloc(pathObj->_elements, pathObj->_max * sizeof(CGPathElementInternal));
-    }
-
-    for (unsigned i = 0; i < copyObj->_count; i++) {
-        pathObj->_elements[pathObj->_count] = copyObj->_elements[i];
-        pathObj->_count++;
-        CGPathElementInternal* c = &pathObj->_elements[pathObj->_count];
-
-        if (m) {
-            switch (c->type) {
-                case kCGPathElementMoveToPoint:
-                case kCGPathElementAddLineToPoint:
-                    c->points[0] = CGPointApplyAffineTransform(c->points[0], *m);
-                    break;
-                default:
-                    // Append the path anyway, without transforming.
-                    break;
-            }
-        }
-    }
-
-    // we need to re init the elements to setup the pointers propertly
-    for (int i = 0; i < path->_count; i++) {
-        CGPathElementInternal* element = &path->_elements[i];
-        element->init();
-    }
+    // Close the path we're adding and open the path we need to add to.
+    toAdd->closePath();
+    path->preparePathForEditing();
+    FAIL_FAST_IF_FAILED(toAdd->_impl.pathGeometry->Stream(path->_impl.geometrySink.Get()));
 }
 
 /**
  @Status Interoperable
 */
-void CGPathAddEllipseInRect(CGMutablePathRef path, const CGAffineTransform* m, CGRect rect) {
+void CGPathAddEllipseInRect(CGMutablePathRef path, const CGAffineTransform* transform, CGRect rect) {
```

Comment:
>CGPathAddEllipseInRect [](start = 5, length = 22)

This is not for this review:
I wonder if we can use path combining, or somehow stream an ellipse geometry into the path geometry.